tests/bin/sessions.test.ts

```diff
@@ -1,215 +1,321 @@
+  }, global.maxTimeout);
+  afterAll(async () => {
+    await pkAgentClose();
+  });
+  beforeEach(async () => {
     dataDir = await fs.promises.mkdtemp(
       path.join(os.tmpdir(), 'polykey-test-'),
     );
-    nodePath = path.join(dataDir, 'keynode');
-    command = ['vaults', 'list', '-np', nodePath];
-    passwordFile = path.join(dataDir, 'passwordFile');
-    sessionFile = path.join(nodePath, 'token');
-    await fs.promises.writeFile(passwordFile, 'password');
-    polykeyAgent = await PolykeyAgent.createPolykeyAgent({
-      password: 'password',
-      nodePath: nodePath,
-      logger: logger,
+    // Invalidate all active sessions
+    await testBinUtils.pkStdio(
+      ['agent', 'lock'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+        PK_PASSWORD: global.binAgentPassword,
+      },
+      global.binAgentDir,
+    );
+  });
+  afterEach(async () => {
+    await fs.promises.rm(dataDir, {
+      force: true,
+      recursive: true,
     });
-    await polykeyAgent.vaultManager.createVault(vaultName);
   });
-
-  afterAll(async () => {
-    await polykeyAgent.stop();
-    await polykeyAgent.destroy();
-    await fs.promises.rmdir(dataDir, { recursive: true });
+  test('processes should store session token in session file', async () => {
+    // Run command to set token
+    let exitCode, stdout;
+    ({ exitCode, stdout } = await testBinUtils.pkStdio(
+      ['agent', 'status', '--verbose'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+        PK_PASSWORD: global.binAgentPassword,
+      },
+      global.binAgentDir,
+    ));
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
+    await sleep(1100);
+    // Try again without password
+    ({ exitCode, stdout } = await testBinUtils.pkStdio(
+      ['agent', 'status', '--verbose'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+      },
+      global.binAgentDir,
+    ));
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
   });
-
-  test('Process should store session token in session file', async () => {
-    // Agent has not been unlocked yet
-    const result = await testUtils.pkStdio(
-      command,
-      { PK_PASSWORD: 'password' },
-      dataDir,
-    );
-    expect(result.exitCode).toBe(0);
-    expect(result.stdout).toContain(vaultName);
-
-    const buff = await fs.promises.readFile(sessionFile);
-    const newToken = buff.toString() as SessionToken;
-    expect(
-      async () => await polykeyAgent.sessionManager.verifyToken(newToken),
-    ).not.toThrow();
+  test('processes should refresh the session token', async () => {
+    const session = await Session.createSession({
+      sessionTokenPath,
+      fs,
+      logger,
+    });
+    // Generate new token
+    await testBinUtils.pkStdio(
+      ['agent', 'unlock'],
+      {
+        PK_NODE_PATH: global.binAgentDir,
+        PK_PASSWORD: global.binAgentPassword,
+      },
+      global.binAgentDir,
+    );
+    await sleep(1100);
```

Comment:
This should not be necessary. It is guaranteed that once the command is done, the session token is changed.

Reply:
I've been playing around with this test (`processes should refresh the session token`), trying to remove all of the `sleep` calls.

Removing all the `sleep` calls from the test gives the following:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken); // TOKEN: undefined
    // Generate new token
    await testBinUtils.pkStdio(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
      },
      global.binAgentDir,
    );

    // await sleep(1100);

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1); // TOKEN: ...ZV0sI1MeZ0HuSco
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkStdio(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2); // TOKEN: ...ZV0sI1MeZ0HuSco
    expect(token1).not.toBe(token2);
  });
```

Strangely, re-adding the `sleep` call immediately after the unlock:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken); // TOKEN: undefined
    // Generate new token
    await testBinUtils.pkStdio(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
      },
      global.binAgentDir,
    );

    await sleep(1100); // uncommented here

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1); // TOKEN: ...ZV0sI1MeZ0HuSco
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkStdio(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2); // TOKEN: ...wM1Bw-NzXdP_880
    expect(token1).not.toBe(token2);
  });
```

That is, despite having no timeout added between the read of `token1` and `token2`, the tokens successfully change after having a timeout immediately after the unlock. This doesn't make a huge amount of sense to me.

Reply:
Running the second version of the test above produces a test execution time of 1600ms according to jest output.

@CMCDragonkai suggested trying out `pkExec` - his thinking was the write to the session write is asynchronous, and may not be being awaited in the session interceptor. Using `pkExec` would wait until the process terminates, which naturally would also require the session token to have been changed.

Doing so does correctly provide the expected change in the session token. However, at least on my machine, this produces a significant increase in test time. i.e. running the following:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken);
    // Generate new token
    await testBinUtils.pkExec(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
      },
      global.binAgentDir,
    );

    // await sleep(1100);

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1);
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkExec(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2);
    expect(token1).not.toBe(token2);
  });
```

Gives a test execution time of 11300ms according to jest:

```
 PASS  tests/bin/sessions.test.ts (26.747 s)
  CLI Sessions
    ✓ processes should refresh the session token (11369 ms)
    ○ skipped processes should store session token in session file
    ○ skipped serial processes should both refresh the session token
    ○ skipped failing processes should unlock the session file
    ○ skipped agent lock should remove the token from the client and delete the token
    ○ skipped cause old sessions to fail when locking all sessions
    ○ skipped unattended calls with invalid auth should fail
    ○ skipped password can be used to authenticate attended calls
    ○ skipped re-prompt for password if unable to authenticate call
    ○ skipped will not prompt for password reauthentication on generic error

Test Suites: 1 passed, 1 total
Tests:       9 skipped, 1 passed, 10 total
Snapshots:   0 total
Time:        26.8 s
Ran all test suites matching /tests\/bin\/sessions.test.ts/i.
GLOBAL TEARDOWN
Destroying global.binAgentDir: /run/user/1000/polykey-test-bin

```

Reply:
The sleeps were added after I kept getting random test failures all over the place - there are likely some that are unecessary though.

Reply:
@CMCDragonkai suggested to try this out https://www.npmjs.com/package/typescript-cached-transpile

Huge improvement when using `pkExec` now:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken);
    // Generate new token
    await testBinUtils.pkExec(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
        TS_NODE_TRANSPILE_ONLY: 'true',
        TS_NODE_COMPILER: 'typescript-cached-transpile',
      },
      global.binAgentDir,
    );

    // await sleep(1100);

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1);
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkExec(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
        TS_NODE_TRANSPILE_ONLY: 'true',
        TS_NODE_COMPILER: 'typescript-cached-transpile',
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2);
    expect(token1).not.toBe(token2);
  });
```

```
 PASS  tests/bin/sessions.test.ts (14.369 s)
  CLI Sessions
    ✓ processes should refresh the session token (2401 ms)
    ○ skipped processes should store session token in session file
    ○ skipped processes should refresh the session token
    ○ skipped serial processes should both refresh the session token
    ○ skipped failing processes should unlock the session file
    ○ skipped agent lock should remove the token from the client and delete the token
    ○ skipped cause old sessions to fail when locking all sessions
    ○ skipped unattended calls with invalid auth should fail
    ○ skipped password can be used to authenticate attended calls
    ○ skipped re-prompt for password if unable to authenticate call
    ○ skipped will not prompt for password reauthentication on generic error

Test Suites: 1 passed, 1 total
Tests:       10 skipped, 1 passed, 11 total
Snapshots:   0 total
Time:        14.421 s
Ran all test suites matching /tests\/bin\/sessions.test.ts/i.
GLOBAL TEARDOWN
Destroying global.binAgentDir: /run/user/1000/polykey-test-bin
```

The first run of the test was 6000ms with these env variables set, but every successive one after was ~2400ms.

Reply:
Hmm... I thought this problem was resolved with usage of `pkExec`, but seems like we still get random failures where the session tokens are still the same after execution of the command (solely in this `processes should refresh the session token` test).

Reply:
Should be resolved with our new `FlowCountInterceptor`.